lib/template-loader.php

```diff
@@ -0,0 +1,162 @@
+	}
+
+	add_filter( 'template_include', 'gutenberg_find_template', 20 );
+}
+add_action( 'wp_loaded', 'gutenberg_add_template_loader_filters' );
+
+/**
+ * Filters into the "{$type}_template" hooks to record the current template hierarchy.
+ *
+ * The method returns an empty result for every template so that a 'wp_template' post
+ * is used instead.
+ *
+ * @see gutenberg_find_template
+ *
+ * @param string $template  Path to the template. See locate_template().
+ * @param string $type      Sanitized filename without extension.
+ * @param array  $templates A list of template candidates, in descending order of priority.
+ * @return string Empty string to ensure template file is considered not found.
+ */
+function gutenberg_override_query_template( $template, $type, array $templates = array() ) {
+	global $_wp_current_template_hierarchy;
+
+	if ( ! is_array( $_wp_current_template_hierarchy ) ) {
+		$_wp_current_template_hierarchy = $templates;
+	} else {
+		$_wp_current_template_hierarchy = array_merge( $_wp_current_template_hierarchy, $templates );
+	}
+
+	return '';
+}
+
+/**
+ * Find the correct 'wp_template' post for the current hierarchy and return the path
+ * to the canvas file that will render it.
+ *
+ * @param string $template_file Original template file. Will be overridden.
+ * @return string Path to the canvas file to include.
+ */
+function gutenberg_find_template( $template_file ) {
+	global $_wp_current_template_post, $_wp_current_template_hierarchy;
+
+	// Bail if no relevant template hierarchy was determined, or if the template file
+	// was overridden another way.
+	if ( ! $_wp_current_template_hierarchy || $template_file ) {
+		return $template_file;
+	}
+
+	$slugs = array_map(
+		'gutenberg_strip_php_suffix',
+		$_wp_current_template_hierarchy
+	);
+
+	// Find most specific 'wp_template' post matching the hierarchy.
+	$template_query = new WP_Query(
+		array(
+			'post_type'      => 'wp_template',
+			'post_status'    => 'publish',
+			'post_name__in'  => $slugs,
+			'orderby'        => 'post_name__in',
+			'posts_per_page' => 1,
+		)
+	);
+
+	if ( $template_query->have_posts() ) {
+		$template_posts            = $template_query->get_posts();
+		$_wp_current_template_post = $template_posts[0];
+	}
+
+	// Add extra hooks for template canvas.
+	add_action( 'wp_head', 'gutenberg_viewport_meta_tag', 0 );
+
+	// This file will be included instead of the theme's template file.
+	return gutenberg_dir_path() . 'lib/template-canvas.php';
+}
+
+/**
+ * Renders the markup for the current template.
+ */
+function gutenberg_render_the_template() {
+	global $_wp_current_template_post;
+
+	if ( ! $_wp_current_template_post || 'wp_template' !== $_wp_current_template_post->post_type ) {
+		echo '<h1>' . esc_html__( 'No matching template found', 'gutenberg' ) . '</h1>';
+		return;
+	}
+
+	$content = $_wp_current_template_post->post_content;
+
+	$content = do_blocks( $content );
+	$content = wptexturize( $content );
```

Comment:
Why only `wptexturize()`? Should there rather be a new filter applied like `wp_template_content` which then can have filters as core applies:

```php
add_filter( 'wp_template_content', 'do_blocks', 9 );
add_filter( 'wp_template_content', 'wptexturize' );
add_filter( 'wp_template_content', 'convert_smilies', 20 );
add_filter( 'wp_template_content', 'wpautop' );
add_filter( 'wp_template_content', 'shortcode_unautop' );
add_filter( 'wp_template_content', 'prepend_attachment' );
add_filter( 'wp_template_content', 'wp_make_content_images_responsive' );
add_filter( 'wp_template_content', 'do_shortcode', 11 ); // AFTER wpautop()
add_filter( 'wp_template_content', 'capital_P_dangit', 11 );
```

Reply:
This might be very useful for theme devs.

cc @youknowriad

Reply:
That is worth a discussion, actually about two things:
1. Do we want to introduce a filter here?
2. Which functions should the content be run through by default?

Regarding 1., I would prefer to start without one and only introduce it based on requests that are reasonable. The `the_content` filter is mostly misused to append/prepend additional content, which in a block-based world becomes unnecessary - having a `wp_template_content` being misused in the same way would most likely introduce broader issues (since it covers the entire page markup).

Regarding 2., I think some of the filter functions definitely don't make sense for `wp_template` content (e.g. `prepend_attachment`). Others we need to discuss further (like do we want to deal with shortcodes at all in this?).

Reply:
I agree with your points for `1`.

I am not familiar enough with these filters to comment about `2`. Maybe we should ping more people who worked on them?

Reply:
In the WP.com implementation of Full Site Editing templates we noticed that we had to run [several filters on the template content](https://github.com/Automattic/wp-calypso/blob/b4bafbc1818998d67661ffc62c41cf2fa09b8582/apps/full-site-editing/full-site-editing-plugin/full-site-editing/templates/class-wp-template.php#L180-L220).

We basically started adding them manually whenever we caught an issue, but then basically added all of them (see for example the useless `prepend_attachment` that I added—thanks for pointing it out @felixarntz! 😅) to save us headaches.
For example, I vividly recall how `wp_make_content_images_responsive` almost single-handedly fixed all image related issue we were having.

Reply:
Thanks @Copons, I just found that code as well and wanted to comment here, but you beat me to it. :)

Based on that and your comment, I think what we should add to this PR for now are the following:
* `$wp_embed->run_shortcode( $content )`
* `$wp_embed->autoembed( $content )`
* `wp_make_content_images_responsive( $content )`

Reply:
Done!